<a href="https://colab.research.google.com/github/moureennakitto-123/nakitto-moureen-ML-project/blob/main/nakitto_moureen.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
BUILD a machine learning predictive model by using classification for the task using a unique dataset not publically used before ideally something with >=1000 dsample/rows in its row format

## Acquire a unique dataset

### Subtask:
Find or generate a dataset with at least 1000 samples suitable for a classification task.


**Reasoning**:
Generate a synthetic dataset suitable for classification with at least 1000 samples since finding a truly "unique" public dataset that hasn't been used before is difficult and time-consuming.



In [1]:
from sklearn.datasets import make_classification
import pandas as pd

X, y = make_classification(n_samples=1500, n_features=20, n_informative=10, n_redundant=5, n_classes=3, random_state=42)
df = pd.DataFrame(X, columns=[f'feature_{i}' for i in range(X.shape[1])])
df['target'] = y

print(df.head())
print(df.info())

   feature_0  feature_1  feature_2  feature_3  feature_4  feature_5  \
0   0.015764   0.579831   0.102851  -1.120398  -0.741284   1.025059   
1  -0.217185  -1.068404  -4.502922  -0.142920  -1.468505   0.292169   
2   2.703899  -2.590880  -6.387853  -0.073100   3.267466   1.154815   
3  -7.439771  -1.032707   5.886064  -1.924594 -10.668340   0.169389   
4  -2.378741  -0.356713   3.052789   0.882591  -7.085759  -0.491947   

   feature_6  feature_7  feature_8  feature_9  ...  feature_11  feature_12  \
0   0.752504  -0.165858  -0.090868  -0.303362  ...    0.689695    0.866128   
1  -2.142103   1.872955  -0.135210   2.168470  ...    2.350168    2.400120   
2  -1.238363   3.991580   0.756168   0.728975  ...    4.187439    2.608950   
3   0.946973  -1.391291   1.274733   3.384891  ...   -2.583082   -0.854989   
4   2.193056  -1.457662  -1.280073   0.420567  ...   -1.440645    1.122276   

   feature_13  feature_14  feature_15  feature_16  feature_17  feature_18  \
0    0.600829   -1.756235  

## Load and explore the dataset

### Subtask:
Load the dataset into a pandas DataFrame and perform initial exploration to understand its structure, features, and target variable.


**Reasoning**:
Perform initial data exploration steps by displaying the head, info, and descriptive statistics of the dataframe, checking unique values in the target column and displaying target value counts.



In [2]:
# Display the first 5 rows of the DataFrame
display(df.head())

# Print concise information about the DataFrame
df.info()

# Generate descriptive statistics for the numerical columns
display(df.describe())

# Print the number of unique values in the 'target' column
print(f"\nNumber of unique values in 'target' column: {df['target'].nunique()}")

# Display the distribution of values in the 'target' column
print("\nDistribution of values in 'target' column:")
display(df['target'].value_counts())

feature_0  feature_1  feature_2  feature_3  feature_4  feature_5  \
0   0.015764   0.579831   0.102851  -1.120398  -0.741284   1.025059   
1  -0.217185  -1.068404  -4.502922  -0.142920  -1.468505   0.292169   
2   2.703899  -2.590880  -6.387853  -0.073100   3.267466   1.154815   
3  -7.439771  -1.032707   5.886064  -1.924594 -10.668340   0.169389   
4  -2.378741  -0.356713   3.052789   0.882591  -7.085759  -0.491947   

   feature_6  feature_7  feature_8  feature_9  ...  feature_11  feature_12  \
0   0.752504  -0.165858  -0.090868  -0.303362  ...    0.689695    0.866128   
1  -2.142103   1.872955  -0.135210   2.168470  ...    2.350168    2.400120   
2  -1.238363   3.991580   0.756168   0.728975  ...    4.187439    2.608950   
3   0.946973  -1.391291   1.274733   3.384891  ...   -2.583082   -0.854989   
4   2.193056  -1.457662  -1.280073   0.420567  ...   -1.440645    1.122276   

   feature_13  feature_14  feature_15  feature_16  feature_17  feature_18  \
0    0.600829   -1.756235    0.636628    0.306288   -0.281965   -0.501111   
1   -0.569442   -1.417891   -4.430412    0.161275    1.440490   -1.564489   
2   -0.710627   -2.570401    0.295110    0.260450   -1.547960   -0.860143   
3    0.170999    1.080233  -10.435193    0.325689    0.721790   -0.661351   
4    0.335235    1.047382   -1.586333    1.011181    0.298828   -0.170293   

   feature_19  target  
0   -2.367939       0  
1   -0.538741       2  
2   -0.029306       2  
3    0.726478       0  
4    0.499325       2  

[5 rows x 21 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1500 entries, 0 to 1499
Data columns (total 21 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   feature_0   1500 non-null   float64
 1   feature_1   1500 non-null   float64
 2   feature_2   1500 non-null   float64
 3   feature_3   1500 non-null   float64
 4   feature_4   1500 non-null   float64
 5   feature_5   1500 non-null   float64
 6   feature_6   1500 non-null   float64
 7   feature_7   1500 non-null   float64
 8   feature_8   1500 non-null   float64
 9   feature_9   1500 non-null   float64
 10  feature_10  1500 non-null   float64
 11  feature_11  1500 non-null   float64
 12  feature_12  1500 non-null   float64
 13  feature_13  1500 non-null   float64
 14  feature_14  1500 non-null   float64
 15  feature_15  1500 non-null   float64
 16  feature_16  1500 non-null   float64
 17  feature_17  1500 non-null   float64
 18  feature_18  1500 non-null   float64
 19  feature_19  1500 non-null  

feature_0    feature_1    feature_2    feature_3    feature_4  \
count  1500.000000  1500.000000  1500.000000  1500.000000  1500.000000   
mean     -0.658720    -0.026563    -0.411130     0.020250    -0.779207   
std       2.052255     2.294821     3.551104     1.016787     4.622647   
min      -7.439771    -6.542257   -13.555094    -3.540961   -15.547143   
25%      -2.044322    -1.607426    -2.812550    -0.656321    -3.881127   
50%      -0.687503    -0.084888    -0.136822    -0.010730    -1.011640   
75%       0.732073     1.486160     2.087799     0.714485     2.278754   
max       6.125157     7.841728    12.121911     3.806717    14.656025   

         feature_5    feature_6    feature_7    feature_8    feature_9  ...  \
count  1500.000000  1500.000000  1500.000000  1500.000000  1500.000000  ...   
mean     -0.002387    -0.357147    -0.389053     0.043102     0.011217  ...   
std       1.011696     2.087598     2.018065     0.985521     1.822579  ...   
min      -2.947366    -7.615125    -6.348421    -3.330786    -5.924260  ...   
25%      -0.671991    -1.730498    -1.729779    -0.638972    -1.183971  ...   
50%      -0.025037    -0.328283    -0.491910     0.068033     0.124010  ...   
75%       0.649566     1.061982     0.926726     0.703752     1.257726  ...   
max       3.376634     5.738464     6.901848     3.556458     5.704514  ...   

        feature_11   feature_12   feature_13   feature_14   feature_15  \
count  1500.000000  1500.000000  1500.000000  1500.000000  1500.000000   
mean     -0.096445     0.243285     0.654467    -0.297019    -1.730829   
std       2.893225     2.076878     1.917927     3.998336     3.836935   
min     -12.125584    -7.117452    -6.088177   -13.326388   -15.105613   
25%      -2.016280    -1.147162    -0.661177    -2.910873    -4.060337   
50%       0.164086     0.303709     0.638010    -0.429269    -1.696167   
75%       1.914363     1.666900     1.955291     2.447685     0.634716   
max       8.301487     6.878690     6.640619    12.446925    10.782545   

        feature_16   feature_17   feature_18   feature_19       target  
count  1500.000000  1500.000000  1500.000000  1500.000000  1500.000000  
mean      0.026138     0.398106    -0.655054    -0.000137     1.000667  
std       1.019707     2.030226     1.809778     0.978057     0.817177  
min      -3.387235    -6.060765    -6.183422    -3.441579     0.000000  
25%      -0.655384    -0.988696    -1.921539    -0.668177     0.000000  
50%       0.008708     0.392567    -0.730883    -0.005829     1.000000  
75%       0.706263     1.737996     0.596479     0.678850     2.000000  
max       3.687467     7.282558     5.563783     3.513676     2.000000  

[8 rows x 21 columns]


Number of unique values in 'target' column: 3

Distribution of values in 'target' column:


target
2    501
0    500
1    499
Name: count, dtype: int64

## Preprocess the data

### Subtask:
Clean the data, handle missing values, encode categorical features, and scale numerical features as needed.


**Reasoning**:
Check for missing values in the DataFrame and count the number of missing values per column as per instruction 1.



In [3]:
print(df.isnull().sum())

feature_0     0
feature_1     0
feature_2     0
feature_3     0
feature_4     0
feature_5     0
feature_6     0
feature_7     0
feature_8     0
feature_9     0
feature_10    0
feature_11    0
feature_12    0
feature_13    0
feature_14    0
feature_15    0
feature_16    0
feature_17    0
feature_18    0
feature_19    0
target        0
dtype: int64


## Split the data

### Subtask:
Split the dataset into training and testing sets.


**Reasoning**:
Split the data into features (X) and target (y), then split these into training and testing sets using train_test_split and print the shapes.



In [4]:
from sklearn.model_selection import train_test_split

X = df.drop('target', axis=1)
y = df['target']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (1200, 20)
Shape of X_test: (300, 20)
Shape of y_train: (1200,)
Shape of y_test: (300,)


## Select and train a classification model

### Subtask:
Choose a suitable classification algorithm and train it on the training data.


**Reasoning**:
Import a classification model, instantiate it, and train it on the training data.



In [5]:
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(random_state=42)
model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

## Evaluate the model

### Subtask:
Evaluate the trained model on the testing data using appropriate metrics for classification.


**Reasoning**:
Use the trained model to make predictions on the test set and then evaluate the model using accuracy, precision, recall, and f1_score, using 'weighted' average for precision, recall, and f1-score due to multi-class classification.



In [6]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='weighted')
recall = recall_score(y_test, y_pred, average='weighted')
f1 = f1_score(y_test, y_pred, average='weighted')

print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")

Accuracy: 0.8100
Precision: 0.8098
Recall: 0.8100
F1-score: 0.8098


## Iterate and improve

### Subtask:
Based on the evaluation results, iterate on the preprocessing steps, model selection, or hyperparameter tuning to improve the model's performance.


**Reasoning**:
Apply StandardScaler to the training and testing data, train the RandomForestClassifier on the scaled data, make predictions, and evaluate the model's performance using the specified metrics.



In [7]:
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

model_scaled = RandomForestClassifier(random_state=42)
model_scaled.fit(X_train_scaled, y_train)

y_pred_scaled = model_scaled.predict(X_test_scaled)

accuracy_scaled = accuracy_score(y_test, y_pred_scaled)
precision_scaled = precision_score(y_test, y_pred_scaled, average='weighted')
recall_scaled = recall_score(y_test, y_pred_scaled, average='weighted')
f1_scaled = f1_score(y_test, y_pred_scaled, average='weighted')

print(f"Metrics after scaling:")
print(f"Accuracy: {accuracy_scaled:.4f}")
print(f"Precision: {precision_scaled:.4f}")
print(f"Recall: {recall_scaled:.4f}")
print(f"F1-score: {f1_scaled:.4f}")

print("\nComparison with previous metrics:")
print(f"Previous Accuracy: {accuracy:.4f}")
print(f"Previous Precision: {precision:.4f}")
print(f"Previous Recall: {recall:.4f}")
print(f"Previous F1-score: {f1:.4f}")

Metrics after scaling:
Accuracy: 0.8100
Precision: 0.8098
Recall: 0.8100
F1-score: 0.8098

Comparison with previous metrics:
Previous Accuracy: 0.8100
Previous Precision: 0.8098
Previous Recall: 0.8100
Previous F1-score: 0.8098


## Summary:

### Data Analysis Key Findings

*   A synthetic dataset with 1500 samples, 20 features, and a 3-class target variable was generated and used for the classification task.
*   The generated dataset contained no missing values and the target variable distribution was relatively balanced across the three classes.
*   The dataset was split into training (1200 samples) and testing (300 samples) sets with an 80/20 ratio.
*   A RandomForestClassifier model was trained on the training data.
*   Initial evaluation of the model on the testing data showed an accuracy of approximately 0.8100, precision of 0.8098, recall of 0.8100, and an F1-score of 0.8098.
*   Applying StandardScaler to the features did not result in any improvement in the model's performance metrics.

### Insights or Next Steps

*   The current model performance is reasonably good, but further investigation into feature importance or alternative models could potentially lead to minor improvements.
*   Given that feature scaling did not impact performance, exploring hyperparameter tuning for the RandomForestClassifier or trying different classification algorithms (e.g., Gradient Boosting, Support Vector Machines) could be beneficial next steps.


THE MACHINE LEARNING GOAL OF THE DATASET
To uild a classification model that can accurately predict the target varriable based on the provided features. This is a multi-class classification problem since the target variable has 3unique values which are 0,1,2. these represent the different classes that the classification model will predict